helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -433,40 +433,116 @@ public void removeDeploymentGroup(final String name) throws DeploymentGroupDoesN
+   *
+   * @param status The status of the deployment group attempting to change hosts.
+   * @return True if it's safe to update the hosts.
+   */
+  private boolean allowHostChange(final DeploymentGroupStatus status) {
+    if (status == null) {
+      // We're in an unknown state. Go hog wild.
+      return true;
+    }
+    return status.getState() != ROLLING_OUT;
+  }
+
+  /**
+   * Determines whether we should trigger a rolling update when deployment group hosts change.
+   *
+   * We want to avoid triggering an automatic rolling update if the most recent rolling update was
+   * triggered manually, and failed.
+   *
+   * @param group The deployment group that is changing hosts.
+   * @param status The status of the aforementioned deployment group.
+   * @return True if we should perform a rolling update.
+   */
+  private boolean updateOnHostChange(final DeploymentGroup group,
+                                     final DeploymentGroupStatus status) {
+    if (status == null) {
+      // We're in an unknown state. Go hog wild.
+      return true;
+    }
+
+    if (group.getRollingUpdateReason() == null) {
+      // The last rolling update didn't fail, so there's no reason to expect this one will.
+      return status.getState() != FAILED;
+    }
+
+    if (group.getRollingUpdateReason() == HOSTS_CHANGED && status.getState() == FAILED) {
+      // The last rolling update failed, but it was triggered by hosts changing. Try again.
+      return true;
+    }
+
+    // The last rolling update didn't fail, so there's no reason to expect this one will.
+    return status.getState() != FAILED;
+  }
+
   @Override
   public void updateDeploymentGroupHosts(final String groupName, final List<String> hosts)
       throws DeploymentGroupDoesNotExistException {
     log.debug("updating deployment-group hosts: name={}", groupName);
     final ZooKeeperClient client = provider.get("updateDeploymentGroupHosts");
     try {
-      Optional<Integer> version = Optional.absent();
+      Optional<Integer> curHostsVersion = Optional.absent();
       List<String> curHosts;
       try {
         // addDeploymentGroup creates Paths.statusDeploymentGroupHosts(name) so it should always
         // exist. If it doesn't, then the DG was (likely) deleted.
-        final Node node = client.getNode(Paths.statusDeploymentGroupHosts(groupName));
-        version = Optional.of(node.getStat().getVersion());
-        curHosts = Json.read(node.getBytes(), new TypeReference<List<String>>() {});
+        final Node chn = client.getNode(Paths.statusDeploymentGroupHosts(groupName));
+        curHostsVersion = Optional.of(chn.getStat().getVersion());
+        curHosts = Json.read(chn.getBytes(), new TypeReference<List<String>>() {});
       } catch (JsonMappingException e) {
         curHosts = Collections.emptyList();
-      } catch (NoNodeException e) {
-        // DG was deleted -- abort.
+      }
+
+      final Node sn = client.getNode(Paths.statusDeploymentGroup(groupName));
```

Comment:
I believe this will throw a `NoNodeException` if the nodes does not exist. You probably want to catch and act on that. The node does not exist --> `status` will be `null` below.